### ISSUE TITLE
Full deployment history?

### ISSUE THREAD

**Issue by daniel-beard** [Microsoft/code-push#334]
- I've noticed that both the CLI and Management SDK default to only showing 50 most recent releases for a deployment. 
- Is it possible to get a full paginated list of all releases for a deployment?
- Alternatively, would it be possible to filter by `binaryAppVersion`?

e.g.

```
code-push deployment history MyApp Release --binaryAppVersion 2.3.1
```
- Current version of CLI: `1.12.6-beta`

**Comment by silhouettes** [comment]
@daniel-beard - that's interesting, we never thought anyone would need more than that. How far back do you need to see? Is it just because you're managing several binary versions in the same deployment (how many, if you don't mind me asking)?

**Comment by daniel-beard** [comment]
Out ideal flow would look like this:

```
MyApp
|--Production
|--Staging
|--Debug / Unstable
```

Where releases would flow from Unstable -> Staging -> Release, so there wouldn't necessarily be a limit of binary versions per deployment. It's trivial to go over the 50 releases for Debug / Unstable & Staging, for Production, 50 *should* be enough for now, but it's easy to see a future where it isn't.

Is this a non-standard way of using code-push?

Additionally, this ticket is fairly important to support the above flow: https://github.com/Microsoft/code-push/issues/267 as the HEAD release on Unstable is often not the release we want to promote to Staging (although Staging HEAD is always the release we want to promote to Production).

Example:
In this example we release periodically to Unstable, but `r2` might have a newer fix we want to promote later. As far as I'm aware, this isn't supported by `code-push` right now? I think if we wanted to promote `r2` after `r4` is released to Unstable, we'd have to release `r2` directly to Staging then promote from there, right?
Additionally in this example, there's probably more than 50 builds in between `r2` and `r4`.

<img width="822" alt="screen shot 2016-10-13 at 10 33 54 am" src="https://cloud.githubusercontent.com/assets/2645146/19359880/ee999e20-9130-11e6-9448-4579fd9073eb.png">


Let me know if you need anything else, or have any suggestions for our flow. Thanks!

**Comment by silhouettes** [comment]
Ahh, I think I understand. We're extremely back-logged at the moment, but I would love to squeeze these two requests in. I'll try my best, but no promises!

**Comment by chendo** [comment]
@daniel-beard We want to do something similar. See #336

**Comment by daniel-beard** [comment]
@silhouettes this is blocked by a backend change right now? Or could this be just implemented via CLI / SDK. I have time to take on any changes required in this repository.

**Comment by silhouettes** [comment]
It's a bit of both... Do you have any thoughts on how we should implement pagination? We could just send the full payload from the server, and then do UI-only pagination in the CLI, although I'm wondering what the perf of that would be

**Comment by daniel-beard** [comment]
Was thinking something along the lines of how octokit handles pagination. Disabled by default, you can make subsequent calls to get other pages. Or you can specify an option like `paginateByDefault`

**Comment by silhouettes** [comment]
So I'd really like to do this, but the back-end change seems non-trivial :( The issue is that even with 50 releases, we're starting to hit ~70kb per deployment, so to support more releases per deployment (e.g. 1000), we'd need to be clever with how we store it in the storage layer itself, so as not to disrupt the performance for other users whenever somebody retrieves theirs. But we definitely have this issue in mind, it's just we currently don't have the resources to tackle it!

**Comment by daniel-beard** [comment]
OK, sounds alright (understood on the resources front :) ). An option for backend is to return stubs or just labels + binaryAppVersions.

**Comment by panarasi** [comment]
@silhouettes With the changes to our backend, would this be possible ? Can we move this to our backlog and see if we can fix this ?

**Comment by silhouettes** [comment]
We can certainly move it into our backlog. It's not unfeasible by any means, we'd just have to think everything through. I don't think our current backend changes would change how this feature is implemented, but it would be good to have a discussion and bounce ideas off people.

**Comment by panarasi** [comment]
Awesome, lets keep this issue open then, and see how we can implement it. Thanks

**Comment by sergey-akhalkov** [comment]
Hi @silhouettes, do I get it right that CodePush server uses [MAX_PACKAGE_HISTORY_LENGTH](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/azure-storage.ts#L154) constant to [keep only last 50 packages per deployment](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/azure-storage.ts#L628)? I think we could prompt user that deployment history has reached 50 packages and that next CodePush release will override first release by new one. To keep all history user will need to perform `code-push deployment maxHistoryLength 1000` command to save `maxHistoryLength` value in the CodePush server database's table and use this value further to show full deployment history. 

In a nutshell - `MAX_PACKAGE_HISTORY_LENGTH` constant will be overwritten by specified by user value (that value will be stored in some database table). In case user does not specified `maxHistoryLength` property CodePush server will use `MAX_PACKAGE_HISTORY_LENGTH = 50` constant's value by default.

Seems like this approach won't affect users who need not full deployment history feature, please let me know if I'm wrong or if you have any other thoughts.

**Comment by silhouettes** [comment]
Hey @sergey-akhalkov, I'm not sure if that is a good idea (not saying it's not, just that I'm not sure), because our current deployment history blobs are already ~70kb with 50 releases. With 1000 releases, they'd be over 1MB each, and I don't know what the performance of that will be, and if it'll affect other users. But I'm open-minded though!

Just brainstorming here about other approaches too:
1. Make the history more 'lightweight' like @daniel-beard says, and then have special operations for fetching the rest of the data. For example, I think most of the size would be coming from the download URL's and diff package map - run `code-push deployment history <app> <deployment> --format json` to see what the server is sending (ignore the metrics, as that is a separate sever route that the CLI is stitching together)
2. 'Compress' the data, for example by storing the URL's more compactly. Possibly we don't need to store the prefix (`https://codepush.blob.core.windows.net/`). Don't know how much space that would save.
3. Implement pagination at the storage layer. This is probably the 'correct' option. One approach would be to create a new blob every time the current blob hits 50 releases. Then, when retrieving the history, fetch the last two blobs in order to piece together the most recent 50 releases.

I would lean towards (2) or (3), but other ideas are welcome too! We need to be careful about any changes that we make here though - a lot of the information like the diff package map and download URL's are being consumed by the acquisition server, which can hit thousands of RPS.

Tagging @iageoghe @panarasi @maxim-pop in case they have any other ideas!

**Comment by sergey-akhalkov** [comment]
Hi @silhouettes, agreed, overwriting `MAX_PACKAGE_HISTORY_LENGTH` isn't good idea due to it requires some actions from developer, it'll also affect both CodePush CLI and CodePush Server. So I suppose it would be better if we'll implement server side only solution.

I've tried to find out what methods will be affected in case of increasing `MAX_PACKAGE_HISTORY_LENGTH` constants value from 50 to 1000.
1. The root method is [getPackageHistoryFromBlob](https://github.com/Microsoft/code-push-service/blob/14c52aec3ba95f667ba2c8e0cdb2a7971e81c1bc/api/script/storage/azure-storage.ts#L873)
2. [getPackageHistory](https://github.com/Microsoft/code-push-service/blob/14c52aec3ba95f667ba2c8e0cdb2a7971e81c1bc/api/script/storage/azure-storage.ts#L659), [getPackageHistoryFromDeploymentKey](https://github.com/Microsoft/code-push-service/blob/14c52aec3ba95f667ba2c8e0cdb2a7971e81c1bc/api/script/storage/azure-storage.ts#L535), [commitPackage](https://github.com/Microsoft/code-push-service/blob/14c52aec3ba95f667ba2c8e0cdb2a7971e81c1bc/api/script/storage/azure-storage.ts#L601) methods are using `getPackageHistoryFromBlob` method.
3. `commitPackage` and `getPackageHistory` methods are used by [management.ts](https://github.com/Microsoft/code-push-service/blob/abe8c2df309d71415e878722254fae2dfe11f8c2/api/script/routes/management.ts) (`release`, `promote`, `rollback` and `history` commands). This methods are triggered by developer, so I don't see here performance impact for real app users.
4.  `getPackageHistoryFromDeploymentKey` is used by [updateCheck ](https://github.com/Microsoft/code-push-service/blob/9a8d76d0ca18a9e233182240586724d8e84592b1/api/script/routes/acquisition.ts#L111) route via [createResponseUsingStorage ](https://github.com/Microsoft/code-push-service/blob/9a8d76d0ca18a9e233182240586724d8e84592b1/api/script/routes/acquisition.ts#L31). But I can see that [response is cached](https://github.com/Microsoft/code-push-service/blob/9a8d76d0ca18a9e233182240586724d8e84592b1/api/script/routes/acquisition.ts#L162).
5. I also can see, that `diff-web-job` also will be [affected](https://github.com/Microsoft/code-push-service/blob/b780cee0fc720f6f220acbaff3849797fbe34661/api/script/utils/package-diffing.ts#L49). But I can see that `getDeploymentHistory` method is called by diff job just ones  while developer releasing or promoting an update.

Summarize:
1. app users won't be affected due to `updateCheck` response method is cached.
2. developers will be affected due to using `release`, `promote`, `rollback` and `history` commands, but I don't sure if it is a problem due to this commands are performed by developers infrequently.
3. diff jobs also will be affected, but I've found this constraints: [first](https://github.com/Microsoft/code-push-service/blob/b780cee0fc720f6f220acbaff3849797fbe34661/api/script/utils/package-diffing.ts#L285) and [second](https://github.com/Microsoft/code-push-service/blob/b780cee0fc720f6f220acbaff3849797fbe34661/api/script/utils/package-diffing.ts#L290)

Solution: 
I like option 3 - implement pagination at the storage layer, but it seems like huge task (I've not heard any other requests about this enhancement, but probably it'll happen in the future) and as for now I can't see simple and reliable way we could implement it (please let me know if you have any ideas).
I think that golden mean is **option 2 - 'compress' the data** and increase `MAX_PACKAGE_HISTORY_LENGTH` value.  Together with `storing the URL's more compactly` ([BlobInfo.url](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/storage.ts#L80), [Package.blobUrl](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/storage.ts#L89) and [Package.manifestBlobUrl](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/storage.ts#L95)) we probably also could remove properties that have default values ([Package.isDisabled](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/storage.ts#L92) and [Package.isMandatory](https://github.com/Microsoft/code-push-service/blob/master/api/script/storage/storage.ts#L93) as example).

**Comment by pniko** [comment]
Hi @daniel-beard, closing this for now as it has been prioritized on our backlog and is a limitation build into our backend currently. As we continue to move things into Mobile Center we will end up working on this item at some point in the future, but it is currently indeterminate.
